src/tags/capture_block.rs

```diff
@@ -40,25 +40,29 @@ impl Renderable for Capture {
 
 pub fn capture_block(
     _tag_name: &str,
-    arguments: &[Token],
-    tokens: &[Element],
+    mut arguments: TagTokenIter,
+    mut tokens: TagBlock,
     options: &LiquidOptions,
 ) -> Result<Box<Renderable>> {
-    let mut args = arguments.iter();
-    let id = match args.next() {
-        Some(&Token::Identifier(ref x)) => x.clone(),
-        x @ Some(_) | x @ None => return Err(unexpected_token_error("identifier", x)),
-    };
-
-    // there should be no trailing tokens after this
-    if let t @ Some(_) = args.next() {
-        return Err(unexpected_token_error("`%}`", t));
-    };
+    let id = arguments
+        .expect_next("Identifier expected")?
+        .expect_identifier()
+        .map_err(TagToken::raise_error)?
+        .to_string();
+
+    // no more arguments should be supplied, trying to supply them is an error
+    if let Some(token) = arguments.next() {
+        return Err(token.raise_error());
+    }
 
-    let t =
-        Template::new(parse(tokens, options).trace_with(|| format!("{{% capture {} %}}", &id))?);
+    let template = Template::new(
+        tokens
+            .parse_all(options)
+            .trace_with(|| format!("{{% capture {} %}}", &id))?,
+    );
 
-    Ok(Box::new(Capture { id, template: t }))
+    tokens.assert_empty();
```

Comment:
Unsure if I missed this, is this in every block?  If not, why is this different?

Reply:
Yes, `assert_empty` is in every block before returning an `Ok` value.

This happened because of the issues I had with ensuring all elements of the block are used. In the beggining, `TagBlock` was passed as a `&mut` to the signature of blocks, which we were both unsatisfied with. That was needed to call `close` at the end of the block. I then implemented a `Drop` trait, soon realizing that was also a no-go because it would fire even in case of an `Err`: this meant that if something went wrong during parsing and the function was interrupted before finishing parsing (returning an `Err`), the `Drop` trait would see that the iterator wasn't exhausting the elements and `panic!` on top of the `Err`.

Before reverting back to the `&mut` method, I started thinking more about the user side. Besides the signature with `&mut` not being very clear about what the user must expect of the structure, if he forgot by mistake to parse some elements the check would be done silently and perhaps he wouldn't even notice or care about it. By forcing users (with documentation) to use `.assert_empty`, they are in some way reminding themselves that they must exhaust the elements of the block. If they don't wish to parse them (for example in `comment_block`) they can just go through all the elements and skip them (that would happen anyway, it was a matter of being done by the user or off-scene).

I found that this way would be more transparent and clear to understand the use of the API. Of course if a better solution comes along, we can always change this at a later point in time.